mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,390 @@
+
+    Attributes:
+        unique_id (int): A unique identifier for this agent.
+        model (Model): A reference to the model instance.
+        self.pos: Position | None = None
+    """
+
+    def __init__(self, unique_id: int, model: Model) -> None:
+        """
+        Create a new agent.
+
+        Args:
+            unique_id (int): A unique identifier for this agent.
+            model (Model): The model instance in which the agent exists.
+        """
+        super().__init__(unique_id, model)
+        self.cell: Cell | None = None
+
+    def move_to(self, cell) -> None:
+        if self.cell is not None:
+            self.cell.remove_agent(self)
+        self.cell = cell
+        cell.add_agent(self)
+
+
+class Cell:
+    __slots__ = [
+        "coordinate",
+        "_connections",
+        "owner",
+        "agents",
+        "capacity",
+        "properties",
+    ]
+
+    def __init__(self, coordinate, owner, capacity: int | None = None) -> None:
+        self.coordinate = coordinate
+        self._connections: list[Cell] = []  # TODO: change to CellCollection?
+        self.agents: dict[
+            Agent, None
+        ] = {}  # TODO:: change to AgentSet or weakrefs? (neither is very performant, )
+        self.capacity = capacity
+        self.properties: dict[str, object] = {}
+        self.owner = owner
+
+    def connect(self, other) -> None:
+        """Connects this cell to another cell."""
+        self._connections.append(other)
+
+    def disconnect(self, other) -> None:
+        """Disconnects this cell from another cell."""
+        self._connections.remove(other)
+
+    def add_agent(self, agent: CellAgent) -> None:
+        """Adds an agent to the cell."""
+        n = len(self.agents)
+
+        if n == 0:
+            self.owner._empties.pop(self.coordinate, None)
+
+        if self.capacity and n >= self.capacity:
+            raise Exception(
+                "ERROR: Cell is full"
+            )  # FIXME we need MESA errors or a proper error
+
+        self.agents[agent] = None
+
+    def remove_agent(self, agent: CellAgent) -> None:
+        """Removes an agent from the cell."""
+        self.agents.pop(agent, None)
+        agent.cell = None
+        if len(self.agents) == 0:
+            self.owner._empties[self.coordinate] = None
+
+    @property
+    def is_empty(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
+    def neighborhood(self, radius=1, include_center=False):
+        return CellCollection(
+            self._neighborhood(radius=radius, include_center=include_center)
+        )
+
+    @cache
+    def _neighborhood(self, radius=1, include_center=False):
```

Comment:
What about reusing some of the optimized algorithms in NetworkX? https://stackoverflow.com/questions/18393842/k-th-order-neighbors-in-graph-python-networkx

Reply:
I tested that. It is slower because it requires us to first generate the networkx graph based on our DiscreteSpace. 

Also, NX uses weighted dijkstra for ego_graphs. In our case there are no weights because we don't represent the edges.

Reply:
The pimplementation](https://networkx.org/documentation/stable/_modules/networkx/generators/ego.html#ego_graph) of `nx.ego_graph` reduces to [`nx.single_source_shortest_path_length`](https://networkx.org/documentation/stable/_modules/networkx/algorithms/shortest_paths/unweighted.html#single_source_shortest_path_length) when it is unweighted.

Reply:
> I tested that. It is slower because it requires us to first generate the networkx graph based on our DiscreteSpace.

What if it is initialized as a NetworkX graph to begin with? But then again, there are cythonized NetworkX alternatives. I know there is a [Rust-based NetworkX clone](https://github.com/Qiskit/rustworkx) that powers Qiskit, the most widely used quantum computing library in Python.

Reply:
Rustworkx is about 20x faster than NetworkX for the weighted Dijkstra according to their [benchmark](https://www.rustworkx.org/benchmarks.html#single-source-shortest-path). I'm currently cooking up for the unweighted version of the benchmark.

Reply:
While developing the proof of principle out of which this PR emerged, I tried basing everything on networkx. It was not fast and the code became horrible. Feel free to experiment with other ideas and let us know what you find.

Reply:
I tested. Rustworkx doesn't implement an equivalent of `nx.single_source_shortest_path_length`, nor does the [`dijkstra_shortest_paths`](https://www.rustworkx.org/apiref/rustworkx.dijkstra_shortest_paths.html#rustworkx.dijkstra_shortest_paths) have a cutoff for radius.
Next, I tried `igraph`, but it is 2x-4x slower than `nx.single_source_shortest_path_length`.

> It was not fast and the code became horrible.

How much slower is it? If it is hypothetically multiplied by 20, would it be faster? This is to consider on whether to invest time in implementing `nx.single_source_shortest_path_length` in Rustworkx, or even cythonizing the said NetworkX function.

Reply:
I don't recall the details of the speed. It was at least an order of magnitude slower. 

However, as important for not pursuing a graph library route any further is that the linked list style data structure now presented here is in my view much cleaner, shorter, more elegant, and easier to use than what I was able to come up with while playing around with something on top of networkx.